### PR TITLE
Upgrade the package

### PR DESCRIPTION
**Pull request by elnurvl**
- PHP is updated to 8.0 ~~(this is the minimum supported version now, as the PHP 7.4 support comes to an end in a few months)~~
- PHPUnit is updated to 9.5
- The code is refactored to be compliant with PHP 8.0 (class properties and methods are now fully typed)
- ~~A constructor was added to the classes for properties of which need to be initialized~~
- Tests were fixed, redundant ones were removed
- `IDENTITY_BAND_PASS` constant was reverted to `PASS` to make it compatible with the `AuthenticateSP` response.
- `www.` were removed from the WSDL URL
- Updated the README file
- Code style was fixed

### CODE REVIEW

**Comment by elnurvl** [comment]
@madmatt , @Firesphere , I would like to know if this repository is abandoned or is there any issue with this PR so you do not merge it? In case it is the latter, I am happy to hear the feedback and apply necessary changes before releasing a major upgrade to the package. In case it is the former one, please just let me know, so I will publish the fork repository as a separate package.

**Comment by Firesphere** [comment]
I'll have to leave this one to @madmatt to comment, as I'm not anywhere near having time to give it a look.

I'm assuming he will be addressing this shortly though.

**Comment by madmatt** [comment]
Hey @elnurvl, this is great - thank you!

Overall it looks great. Ideally I'd want to keep support for PHP 7 until it is EOL'd - either way, any update to PHP compatibility would necessitate a new major version of the module. Perhaps this will help us tag v1.0 finally! ;)

I'll take a look at the rest of the PR shortly, thanks again for this!

Note that there's also a few existing forks of this module, I've chatted to the owner of one of the most often-used ones about merging back some of the changes from that into this base repository to make it more useful for people so I'm hoping to do that at some point as well, but that's lower priority than this.

**Comment by elnurvl** [comment]
@madmatt , the EOL is in 3 months. Adding PHP 7 compatibility would prevent us from utilizing the cool syntax features which make the code safer and nice to have in the first major release of the package.

**Comment by madmatt** [comment]
Hey @elnurvl, per the [Supported Versions page on php.net](https://www.php.net/supported-versions.php), while active support ends for PHP 7.4 in 2 months, it's not EOL (unsupported for security updates) for 14 months.

I think overall I'd prefer to support PHP 7 for as long as it's still supported for security updates, given [Packagist currently shows](https://packagist.org/packages/madmatt/id3global-service/php-stats) that all installs of the package currently occur under PHP 7 (this isn't a good gauge but it's the best language-level stats we have unfortunately). edit: It also looks like [one of the larger forks](https://packagist.org/packages/decrypted/id3global-service/php-stats) that I'd like to eventually merge back in also entirely used on PHP 7.

What do you think it would take to support 7.4 as well? I've had a quick look and I didn't see anything too troubling in your PR that would cause incompatibility with 7.4 at least (I don't care about older versions of PHP, I agree that it's not worth supporting PHP 7.3 even though that's in security support, as that runs out in 3 months time and would require a fair amount of tweaks (removing all scalar type hints which are really useful).

Other than that, the changes look great! I still need to go through in detail, but so far I'm a third of the way through and don't have any other issues at all.

**Comment by elnurvl** [comment]
@madmatt you are right, I confused PHP 7.4 and PHP 7.3. It seems PHP 7.4 brings no significant degradation in the syntax. I have added a support for it. If everything else is fine we can merge the changes and publish the package in no time. I need to use it in our production application as soon as possible.

**Comment by madmatt** [comment]
Hey @elnurvl, I see you're still making a few changes to the PR, do you have any more to make at this point? Ideally I'd only review this all once before merging, thanks!

**Comment by elnurvl** [comment]
@madmatt , I do not think I am going to make any further changes for this PR particularly. However, I have been thinking about merging the `FixedFormatAddress` and `FreeFormatAddress` in a separate PR before releasing the major update. The fields have been defined in a single class in the ID3global documentation, so I think it makes sense to make the package API compliant with the official API. Anyway, this is a topic for another discussion, just wanted to let you know about the idea.

**Comment by madmatt** [comment]
Ok perfect thanks - I'll review this again now.

Regarding the change to addresses, you're talking about the [GlobalAddress](https://www.id3globalsupport.com/Website/content/Sample%20Code/Web%20Dev%20Guide%20HTML/html/8c8f8411-100c-31f7-8542-6025967dd274.htm) object, is that right?

That's a fair point, I'd be happy with that change. Previously their code samples only had the separate types of addresses and you had to specify which type of address it was (fixed or free) as I recall, hence why it was built this way.

I agree with you - I'd prefer this module mimics their API as best as we can in PHP. Ideally we'd just include a Swagger client, but given they don't publish that, this is the best we can do :)

**Comment by elnurvl** [comment]
Yes, I am talking about that. Thanks for clearing up. I am happy we agree on this.

File: README.md
Comment (madmatt) [inline README.md:72]: Per [the documentation](https://www.id3globalsupport.com/Website/content/Sample%20Code/Web%20Dev%20Guide%20HTML/html/4602c283-d390-00e8-bd29-bce0236d8bf2.htm), this should be 'Male', 'Female', 'Unknown', 'Unspecified'. Ideally I'd like this to be a const value I think to make it easier, but I never actually implemented that, so let's not change it here.

I've created #11 to resolve this later.

File: composer.json
Comment (madmatt) [inline composer.json:15]: Just as a side-note, I have an unfortunate requirement to still use phpunit 5 which is hopefully going away soon, so I may adjust this after the merge to require `phpunit/phpunit: *` if it fails to install. If so, I'll add a note to the README that the tests will only work on phpunit >9 despite what composer says.

File: composer.json
Comment (madmatt) [inline composer.json]: I'd prefer to leave the old namespace (of `ID3Global\Tests\*`) please. From what I've seen, this is more common for library type packages, and the top-level `Tests` namespace is more likely to be polluted with other project code etc.

File: README.md
Comment (madmatt) [inline README.md]: See below comment on the `GlobalAuthenticationService` class, that will impact this as well if you agree to change it :)

File: src/Service/GlobalAuthenticationService.php
Comment (madmatt) [inline src/Service/GlobalAuthenticationService.php]: This now forces you to specify the identity during service construction, as well as a gateway. The gateway is less of an issue - one service should generally be akin to one gateway (and that's the same in this case), however I'm less okay with passing `Identity` and `$profileId` in here.

In particular, the `IGlobalAuthenticate` endpoint port in the WSDL file defines the following functions: `AuthenticateSP`, `IncrementalVerification`, `AuthenticateMP`, `ForceOverride` and `AddressLookup`. _Theoretically_, at least in my head, this service class might reasonably implement methods for all of these (realistically I think only `AuthenticateMP` makes sense). I'd rather not tie the constructor to what is required for just one of these WSDL ports - hopefully that makes sense?

I'd prefer if we just use the setters as necessary prior to calling ->verifyIdentity(), it feels easier to extend in future that way.

Note: Adjusting this will mean the README will have to adjust too.

Happy to hear your thoughts on this, it's probably my biggest piece of feedback :)

File: src/Identity/Identity.php
Comment (madmatt) [inline src/Identity/Identity.php]: I'm intrigued by this change - by any chance have you discussed with GPG how the 'BandText' field is handled? For my use case, the BandText we get back is the old value (`Name, Address & DOB Match`. But their documentation alternatively states `PASS` or `Pass`.

I note there are other forks of this module that also change this to PASS, so it might be an idiosyncrasy of the specific implementation GPG has done for me - but if so, we might have to expose this as a configuration option to describe what a 'passing' BandText should be.

Nothing for you to change here either way, I'm just interested to see what you've seen yourself.

Thanks!

**Review by madmatt — CHANGES_REQUESTED** [review]
Hey @elnurvl, have completed my review and overall this looks excellent, thanks so much for the changes! I've made a few inline comments, but mostly these are queries for you rather than explicit requests for code changes.

Thanks again for working on this! Let me know if you oppose any of the above or have other thoughts.

File: README.md
Comment (madmatt) [reply README.md:72]: To be clear: There's nothing for you to specifically do here, this is just a note for you :)

File: src/Service/GlobalAuthenticationService.php
Comment (elnurvl) [reply src/Service/GlobalAuthenticationService.php]: I agree. Initially I wanted to take off these properties and make the `verifyIdentity` method accept them as arguments, since they are not part of the state, rather inputs for the service functions. I just did not want to make much change in the class structure without consulting. I am happy to see you think that way now.

File: src/Identity/Identity.php
Comment (elnurvl) [reply src/Identity/Identity.php]: By default, it is `PASS` for successful responses. But you are free to change it to any string in the ID3global's admin page.

Comment (elnurvl) [reply src/Identity/Identity.php]: I do not think this should be a part of the package, since this is a user-defined property not part of the API spec. Clients of the package can use whatever they want to compare the responded band text against.

File: composer.json
Comment (elnurvl) [reply composer.json:15]: I did not get the reason of the requirement. Why tests should fail if they run in PHP ^7.4|^8.0 environment?

File: README.md
Comment (elnurvl) [reply README.md:72]: I thought to use an enumeration for this property, but it would require installing a library (such as `myclabs/php-enum`) unless the package upgraded to PHP 8.1.

File: src/Identity/Identity.php
Comment (elnurvl) [reply src/Identity/Identity.php]: Now I see it may be useful for clients to test their code when they use `GlobalAuthenticationGatewayFake`.

Comment (elnurvl) [reply src/Identity/Identity.php]: @madmatt , I have moved the band text values to the `GlobalAuthenticationGatewayFake` class. Individual users of the package can(should) define their own constants according to the set rules and use them instead to check the response status.

File: composer.json
Comment (madmatt) [reply composer.json:15]: All good, I may update this later. Nothing for you to do now, I'm just stuck on an older version of phpunit for one project where I use this module myself. Will make sure CI etc are configured to use the latest versions of phpunit per your update.

**Comment by madmatt** [comment]
Hey @elnurvl, thanks for your updates. Sorry, I fear that I may have confused you with some of my comments, my apologies! I didn't mean for you to remove the band texts from `Identity`, part of the point of this module is that it should be able to give you a straight 'yes' or 'no' when asked to verify an identity.

What I'll do is review your final changes, merge it in, and then follow up with a separate PR reverting a couple of the changes back to what I meant, rather than make you do any more work on this given it's my fault about the confusion! I'll try to do that this afternoon my time (next ~8hrs or so) as I know you're wanting to use this ASAP in production yourself.

Thanks again for you work on this PR!

**Comment by elnurvl** [comment]
@madmatt , thank you! I am looking forward to see the v1.0!

On the band text, as I mentioned before it does not have a standard value defined by ID3global. For every user it may be different, that is why I think the library should not provide a default band text value for users to check against. It only makes sense in a testing scenario, where one uses the fake gateway instead of a real ID3global API. Because in the fake gateway the band text for the successful response is defined to be `PASS`. For a production project, users should define their band text values in their code rather than relying on this library.

Another note, can you please merge the `FixedFormatAddress` and `FreeFormatAddress` before making your changes, so I would not have to submit another PR for this. I am in a time limit, so doing it myself may delay the release of v1.0

**Comment by madmatt** [comment]
Alright, I've made some changes, the biggest ones are to the verifyIdentity() method signature. I think there's a difference between the profile ID, profile versions values, and the Identity and customer reference values. The profile ones are generally going to be the same on every call, whereas the Identity and customer reference values will change every time you call the service. So my thinking is to split these into separate setter methods that you should call first.

e.g. instead of `$service->verifyIdentity($identity, $profileId, $profileVersion, $customerRef);`, it's now `$service->setProfileId('abc')->setProfileVersion(2)->verifyIdentity($identity, $customerRef);`. This way it's more flexible if you want to verify multiple identities against the same profile (which was one of my original use cases for this module).

Sorry for this late change @elnurvl.

I'll merge your change, then add my changes on top of that, and then will release v1.0.0-beta1 which you can install via composer if you like (you'll need to add `minimum-stability: dev, prefer-stable: true` to your root composer.json). If you're happy with this, I can tag up a stable v1.0.0 version as well (at that point you can remove the min-stability dev).

Hope my reasoning here makes sense to you! I've confirmed the tests still pass on PHP 7.4 and 8.

**Comment by madmatt** [comment]
Regarding the needed address changes, I'm not sure of the best way to approach this. Should we leave the two existing classes as they are and create a new `GlobalAddress` class that just encompasses all the fields maybe? This will retain backwards-compatibility for at least addresses for what that's worth.

My bigger question is that ID3global, as far as I can tell, requires one or other of these address formats to be filled out, and it varies based on country which it requires. As far as I can tell, ID3global **won't** let you submit an address with, for example, `Line1 = '123 Fake St', Country = 'NZ'`. In other words, you need to _either_ use the fixed-field address, or the free-format address, but you can't use both. In that case, doesn't it make sense to provide separate classes instead of combining them?

I admit to not having tried this extensively with their API, so let me know if you have different experiences here.

**Comment by madmatt** [comment]
@elnurvl, new version [1.0.0-beta1](https://github.com/madmatt/id3global-service/releases/tag/1.0.0-beta1) is now released. I've added some release notes there to help with upgrading, and will also add these to the eventual 1.0.0 release.

Let me know if this all makes sense to you, and if it works fine, I can release the stable 1.0.0.

**Comment by elnurvl** [comment]
@madmatt , your changes make sense to me.

The issue with the current structure is you loose the ability of resolving the address type when you try to access it later from the `Identity` object. Unfortunately, this restriction stems from the language itself. However, if ID3global has a restriction on the address fields like you described above, I would prefer to keep different formats separated, regardless. I expected this behavior would be documented in the Online WSDL viewer: every field is marked as nullable and optional there.

Thanks for your collaboration!

**Comment by madmatt** [comment]
Hi @elnurvl, no problem - thank you for your pull request and being understanding about all the requested changes and last minute amendments. Hopefully this all works well for you now!

I've tagged the [1.0.0 stable release](https://github.com/madmatt/id3global-service/releases/tag/1.0.0) now, if you have any issues please feel free to raise issues. If you don't hear back from me, feel free to email my work address (matt at silverstripe dot com) - my GitHub notifications are an absolute dumpster fire and I often miss things, sorry!

I'll be upgrading to this version soon too, but I won't be in a position to test it against production information for some time.

**Comment by elnurvl** [comment]
@madmatt , thank you very much! I already started using the stable release. I have found out some minor issues and submitted a fix PR for those.

**Pull request merged**
